rmw_fastrtps_cpp/src/register_node.cpp

```diff
@@ -0,0 +1,217 @@
+          RCUTILS_STRINGIFY(__function__) "' failed.\n");
+      }
+    });
+  if (!publisher) {
+    return RMW_RET_BAD_ALLOC;
+  }
+
+  // If we would have support for keyed topics, this could be KEEP_LAST and depth 1.
+  qos.history = RMW_QOS_POLICY_HISTORY_KEEP_ALL;
+  std::unique_ptr<rmw_subscription_t, std::function<void(rmw_subscription_t *)>>
+  subscription(
+    rmw_fastrtps_cpp::create_subscription(
+      participant_info.get(),
+      rosidl_typesupport_cpp::get_message_type_support_handle<ParticipantEntitiesInfo>(),
+      "_participant_info",
+      &qos,
+      &subscription_options,
+      false,  // our fastrtps typesupport doesn't support keyed topics
+      true),
+    [&](rmw_subscription_t * sub) {
+      if (RMW_RET_OK != rmw_fastrtps_shared_cpp::destroy_subscription(
+        eprosima_fastrtps_identifier,
+        participant_info.get(),
+        sub))
+      {
+        fprintf(
+          stderr, "Failed to destroy subscription after function: '"
+          RCUTILS_STRINGIFY(__function__) "' failed.\n");
+      }
+    });
+  if (!subscription) {
+    return RMW_RET_BAD_ALLOC;
+  }
+
+  common_context->gid = rmw_fastrtps_shared_cpp::create_rmw_gid(
+    eprosima_fastrtps_identifier, participant_info->participant->getGuid());
+  common_context->pub = publisher.get();
+  common_context->sub = subscription.get();
+
+  context->impl->common = common_context.get();
+  context->impl->participant_info = participant_info.get();
+
+  if (RMW_RET_OK != rmw_fastrtps_cpp::run_listener_thread(context)) {
+    return RMW_RET_ERROR;
+  }
+  common_context->graph_cache.add_participant(common_context->gid);
+
+  publisher.release();
+  subscription.release();
+  common_context.release();
+  participant_info.release();
+  return RMW_RET_OK;
+}
+
+rmw_ret_t
+rmw_fastrtps_cpp::register_node(rmw_context_t * context)
+{
+  assert(context);
+  assert(context->impl);
+
+  std::lock_guard<std::mutex> guard(context->impl->mutex);
+
+  if (!context->impl->count) {
+    rmw_ret_t ret = init_context_impl(context);
+    if (RMW_RET_OK != ret) {
+      return ret;
+    }
+  }
+  context->impl->count++;
+  return RMW_RET_OK;
+}
+
+rmw_ret_t
+rmw_fastrtps_cpp::unregister_node(rmw_context_t * context)
+{
+  assert(context);
+  assert(context->impl);
+
+  std::lock_guard<std::mutex> guard(context->impl->mutex);
+  if (0u == --context->impl->count) {
+    rmw_ret_t ret;
+    ret = rmw_fastrtps_cpp::join_listener_thread(context);
+    if (RMW_RET_OK != ret) {
+      return ret;
+    }
+
+    auto common_context = static_cast<rmw_dds_common::Context *>(context->impl->common);
+    auto participant_info = static_cast<CustomParticipantInfo *>(context->impl->participant_info);
+
+    common_context->graph_cache.remove_participant(common_context->gid);
+    ret = rmw_fastrtps_shared_cpp::destroy_subscription(
+      eprosima_fastrtps_identifier,
+      participant_info,
+      common_context->sub);
+    if (RMW_RET_OK != ret) {
+      return ret;
```

Comment:
Should any attempt be made to destroy the other entities before returning?

Reply:
That's a good question. I will modify the code, so the rest of the cleanup is executed.

Reply:
The one I will left a return, is `join_listener_thread`. If it fails, there might be still a thread running that's using the other objects, so destroying them might lead to UB.

Reply:
done